extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/InlineSchemasAvroBytesDecoder.java

```diff
@@ -0,0 +1,113 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.data.input.avro;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.metamx.common.IAE;
+import com.metamx.common.logger.Logger;
+import com.metamx.common.parsers.ParseException;
+import io.druid.guice.annotations.Json;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.util.ByteBufferInputStream;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ */
+public class InlineSchemasAvroBytesDecoder implements AvroBytesDecoder
+{
+  private static final Logger logger = new Logger(InlineSchemasAvroBytesDecoder.class);
+
+  private final Map<Integer, Schema> schemaObjs;
+  private final Map<String, Map<String, Object>> schemas;
+
+  @JsonCreator
+  public InlineSchemasAvroBytesDecoder(
+      @JacksonInject @Json ObjectMapper mapper,
+      @JsonProperty("schemas") Map<String, Map<String, Object>> schemas
+  ) throws Exception
+  {
+    Preconditions.checkArgument(
+        schemas != null && schemas.size() > 0,
+        "at least one schema must be provided in schemas attribute"
+    );
+
+    this.schemas = schemas;
+
+    schemaObjs = new HashMap<>(schemas.size());
+    for (Map.Entry<String, Map<String, Object>> e : schemas.entrySet()) {
+
+      int id = Integer.parseInt(e.getKey());
+
+      Map<String, Object> schema = e.getValue();
+      String schemaStr = mapper.writeValueAsString(schema);;
+
+      logger.info("Schema string [%s] = [%s]", id, schemaStr);
+      schemaObjs.put(id, new Schema.Parser().parse(schemaStr));
+    }
+  }
+
+  @VisibleForTesting
+  public InlineSchemasAvroBytesDecoder(
+      Map<Integer, Schema> schemaObjs
+  )
+  {
+    this.schemaObjs = schemaObjs;
+    this.schemas = null;
+  }
+
+  @JsonProperty
+  public Map<String, Map<String, Object>> getSchemas()
+  {
+    return schemas;
+  }
+
+  @Override
+  public GenericRecord parse(ByteBuffer bytes)
+  {
+    int schemaId = bytes.getInt();
```

Comment:
Is having a schema id in the first 4 bytes of a message a standard Avro thing? Or is this something specific to our implementation?